yktrace/src/sir.rs

```diff
@@ -31,36 +31,42 @@ pub struct Sir<'m> {
     /// Maps a codegen unit hash to a `(section-name, header, header-size)` tuple.
     hdrs: FxHashMap<CguHash, (String, SirHeader, SirOffset)>,
     /// The current executable's ELF information.
-    exe_obj: object::File<'m>
+    exe_obj: object::File<'m>,
+    /// Section cache to avoid expensive `object::File::section_by_name()` calls.
+    sec_cache: FxHashMap<String, &'m [u8]>
```

Comment:
Can this be `FxHashMap<&'m [u8], &'m [u8]>` or similar?

Reply:
I'm not sure what you mean. We need to look up the section data by symbol name (i.e. a string).

Reply:
Ah, you mean as an `&str`?

Reply:
Yes, but an *owned* string or a *borrowed* string for the key? The value is borrowed (which is cheap), but not the key.

Reply:
Yeah, gotcha. I'll try it.

Reply:
So I think you mean something like this:
```
diff --git a/yktrace/src/sir.rs b/yktrace/src/sir.rs
index a25e545..a355e04 100644
--- a/yktrace/src/sir.rs
+++ b/yktrace/src/sir.rs
@@ -33,7 +33,7 @@ pub struct Sir<'m> {
     /// The current executable's ELF information.
     exe_obj: object::File<'m>,
     /// Section cache to avoid expensive `object::File::section_by_name()` calls.
-    sec_cache: FxHashMap<String, &'m [u8]>
+    sec_cache: FxHashMap<&'m str, &'m [u8]>
 }

 impl<'m> Sir<'m> {
@@ -45,7 +45,7 @@ impl<'m> Sir<'m> {
         for sec in exe_obj.sections() {
             let sec_name = sec.name().unwrap();
             let sec_data = sec.data().unwrap();
-            sec_cache.insert(sec_name.to_owned(), sec_data);
+            sec_cache.insert(sec_name, sec_data);
             if sec_name.starts_with(ykpack::SIR_SECTION_PREFIX) {
                 let mut curs = Cursor::new(sec_data);
                 let mut dec = Decoder::from(&mut curs);
```

This gives:
```
error[E0597]: `exe_obj` does not live long enough
  --> yktrace/src/sir.rs:45:20
   |
39 | impl<'m> Sir<'m> {
   |      -- lifetime `'m` defined here
...
45 |         for sec in exe_obj.sections() {
   |                    ^^^^^^^-----------
   |                    |
   |                    borrowed value does not live long enough
   |                    argument requires that `exe_obj` is borrowed for `'m`
...
66 |     }
   |     - `exe_obj` dropped here while still borrowed
   
error[E0597]: `sec` does not live long enough
  --> yktrace/src/sir.rs:46:28
   |
39 | impl<'m> Sir<'m> {
   |      -- lifetime `'m` defined here
...
46 |             let sec_name = sec.name().unwrap();
   |                            ^^^-------
   |                            |
   |                            borrowed value does not live long enough
   |                            argument requires that `sec` is borrowed for `'m`
...
60 |         }
   |         - `sec` dropped here while still borrowed
   
error[E0505]: cannot move out of `exe_obj` because it is borrowed
  --> yktrace/src/sir.rs:63:13
   |
39 | impl<'m> Sir<'m> {
   |      -- lifetime `'m` defined here
...
45 |         for sec in exe_obj.sections() {
   |                    ------------------
   |                    |
   |                    borrow of `exe_obj` occurs here
   |                    argument requires that `exe_obj` is borrowed for `'m`
...
63 |             exe_obj,
   |             ^^^^^^^ move out of `exe_obj` occurs here
```

So the section name dies with the section.

I briefly tried making the section hang around longer, but quickly got into lifetime spaghetti.

Let's take the 2x and move on? The clone of the section names happens only once at initialisation time, so I doubt we'd have got a big speed up anyway.

Reply:
OK.